### PR TITLE
Refactor mform validation

### PR DESCRIPTION
**Pull request by nhoobin**

### CODE REVIEW

File: classes/form/enrolkey_signup_form.php [hunk 1]
```diff
@@ -78,49 +78,66 @@ public function definition() {
...
+        $signuptoken = $data['signup_token'];
 
-        $token = $data['signup_token'];
+        if ($signuptoken !== '') {
+            // For any case where the token is populated, perform a lookup.
+            $tokenisvalid = $this->check_database_for_enrolkey($signuptoken);
 
-        if (!empty($token)) {
-            $selfenrolinstance = false;
+            if ($tokenisvalid === false) {
```
Comment (Peterburnett) [inline classes/form/enrolkey_signup_form.php:89]: could mildy simplify to !$tokenisvaldid

File: classes/form/enrolkey_signup_form.php [hunk 2]
```diff
@@ -78,49 +78,66 @@ public function definition() {
...
+        return $errors;
+    }
+
+    /**
+     * Checks the enrolment records for any matching self enrolment key.
+     *
+     * @param $token Returns true on success. False on failure.
+     * @return bool
+     */
+    private function check_database_for_enrolkey($token) {
```
Comment (Peterburnett) [inline classes/form/enrolkey_signup_form.php]: I think this should probably be renamed to check_database_for_signuptoken as the rest of the class all references it as signuptoken

File: classes/form/enrolkey_signup_form.php [hunk 3]
```diff
@@ -58,7 +58,7 @@ public function definition() {
         $mform->insertElementBefore($element, 'email');
 
         if ($this->signup_token_required()) {
-            $mform->addRule('signup_token', get_string('signup_missing', 'auth_enrolkey'), 'required', null, 'server');
+            $mform->addRule('signup_token', get_string('signup_missing', 'auth_enrolkey'), 'required', null, 'client');
```
Comment (Peterburnett) [inline classes/form/enrolkey_signup_form.php:61]: Why was this moved clientside?

Comment (nhoobin) [reply classes/form/enrolkey_signup_form.php:61]: This enables the element to display a frontend message when the field is empty, it aligns with the rest of the elements on the sign-up page.
Validation is still performed in the validation() function.

File: classes/form/enrolkey_signup_form.php (hunk 2 above)
Comment (nhoobin) [reply classes/form/enrolkey_signup_form.php]: Cheers, good point.

File: classes/form/enrolkey_signup_form.php (hunk 1 above)
Comment (nhoobin) [reply classes/form/enrolkey_signup_form.php:89]: I feel that this check is more verbose, the reason why this patch exists was to get away from !empty() evaluating the character '0' as true in the earlier statement. I'd like to be able to read what is the exact condition.

File: classes/form/enrolkey_signup_form.php (hunk 3 above)
Comment (nhoobin) [reply classes/form/enrolkey_signup_form.php:61]: An example from core,

https://github.com/moodle/moodle/blob/38abfb6a01888145878e7175e0431a156f6f1078/login/signup_form.php#L51
